### PR TITLE
Implement VanishState

### PR DESCRIPTION
**Pull request by gabizou**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/2405) | **Sponge**

This implements the VanishState with backwards compatibility utilizing the new state entirely.

### CODE REVIEW

File: src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java [hunk 1]
```diff
@@ -324,8 +324,8 @@ public DataContainer toContainer() {
     @Override
     public boolean canSee(final org.spongepowered.api.entity.Entity entity) {
         // note: this implementation will be changing with contextual data
-        final Optional<Boolean> optional = entity.get(Keys.VANISH);
-        return (!optional.isPresent() || !optional.get()) && !((VanishableBridge) entity).bridge$isVanished();
+        final Optional<VanishState> optional = entity.get(Keys.VANISH_STATE);
+        return (!optional.isPresent() || !optional.map(VanishState::isVanished).orElse(true)) && !((VanishableBridge) entity).bridge$vanishState().isVanished();
```
Comment (ImMorpheus) [inline src/mixins/java/org/spongepowered/common/mixin/api/minecraft/world/entity/EntityMixin_API.java]: This should be `orElse(false)` and the `isPresent` check can be removed

File: src/mixins/java/org/spongepowered/common/mixin/core/world/entity/LivingEntityMixin.java [hunk 2]
```diff
@@ -670,7 +670,7 @@ public boolean hurt(final DamageSource source, final float amount) {
      */
     @Inject(method = "isPickable", at = @At("HEAD"), cancellable = true)
     private void impl$ifVanishedDoNotPick(CallbackInfoReturnable<Boolean> cir) {
-        if (this.bridge$isVanished() && this.bridge$isVanishIgnoresCollision()) {
+        if (!this.bridge$vanishState().ignoresCollisions()) {
```
Comment (ImMorpheus) [inline src/mixins/java/org/spongepowered/common/mixin/core/world/entity/LivingEntityMixin.java]: Is the additional `!` intended ?

File: src/mixins/java/org/spongepowered/common/mixin/core/world/entity/player/PlayerMixin.java [hunk 3]
```diff
@@ -170,7 +170,7 @@
 
     @Override
     public boolean bridge$affectsSpawning() {
-        return this.impl$affectsSpawning && !this.shadow$isSpectator() && !this.bridge$isVanishPreventsTargeting();
+        return this.impl$affectsSpawning && (!this.shadow$isSpectator() || !this.bridge$vanishState().untargetable());
```
Comment (ImMorpheus) [inline src/mixins/java/org/spongepowered/common/mixin/core/world/entity/player/PlayerMixin.java]: Why was the condition changed ?

nit: Should be use a separate key for "affect spawning" ?

File: src/main/java/org/spongepowered/common/entity/effect/VanishedVanishState.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.entity.effect;
+
+import com.google.common.base.Objects;
+import org.spongepowered.api.effect.VanishState;
+
+public class VanishedVanishState implements VanishState {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/entity/effect/VanishedVanishState.java]: `final`

Also `SpongeVanishState`

File: src/main/java/org/spongepowered/common/entity/effect/VanishStateFactory.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.entity.effect;
+
+import org.spongepowered.api.effect.VanishState;
+
+public final class VanishStateFactory implements VanishState.Factory {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/entity/effect/VanishStateFactory.java]: Factory should be a subclass in `SpongeVanishState` to remove class bloat.

File: src/main/java/org/spongepowered/common/entity/effect/VisibleVanishState.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.entity.effect;
+
+import org.spongepowered.api.effect.VanishState;
+
+public final class VisibleVanishState implements VanishState {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/common/entity/effect/VisibleVanishState.java]: Could you not remove this and simply set the flags on the previous impl? Is there truly an issue with adjusting the state off of this one (which produces a new state?)

File: testplugins/src/main/java/org/spongepowered/test/vanish/VanishTest.java [hunk 7]
```diff
@@ -32,13 +32,17 @@
...
 import org.spongepowered.api.entity.Entity;
 import org.spongepowered.api.event.Listener;
 import org.spongepowered.api.event.lifecycle.RegisterCommandEvent;
 import org.spongepowered.plugin.PluginContainer;
 import org.spongepowered.plugin.builtin.jvm.Plugin;
 
+import java.util.Collection;
+
 @Plugin("vanish-test")
 public class VanishTest {
```
Comment (Zidane) [inline testplugins/src/main/java/org/spongepowered/test/vanish/VanishTest.java]: `final`

File: src/mixins/java/org/spongepowered/common/mixin/core/world/entity/EntityMixin.java [hunk 8]
```diff
@@ -1094,7 +1054,7 @@ public void stopRiding() {
         final Level instance, final ParticleOptions particleOptions, final double xCoord, final double yCoord,
         final double zCoord, final double xOffset, final double yOffset, final double zOffset
     ) {
-        if (!this.impl$isVanished) {
+        if (!this.bridge$vanishState().createsParticles()) {
```
Comment (ImMorpheus) [inline src/mixins/java/org/spongepowered/common/mixin/core/world/entity/EntityMixin.java]:
```suggestion
        if (this.bridge$vanishState().createsParticles()) {
```

File: src/main/java/org/spongepowered/common/entity/effect/VisibleVanishState.java (hunk 6 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/common/entity/effect/VisibleVanishState.java]: It just makes the Boolean logic a little unwieldy, so:
```java
    public VanishState affectMonsterSpawning(boolean affectsMonsterSpawning) {
        if (!this.vanished) {
            return this;
        }
        if (this.affectsSpawning == affectsMonsterSpawning) {
            return this;
        }
        return new SpongeVanishState(
            true,
            this.ignoresCollisions,
            this.untargetable,
            affectsMonsterSpawning,
            this.createsSounds,
            this.createsParticles
        );
    }
```
would be the odd thing of keeping the tristate (are we vanished) in the method, whereas the "Oh, we're a completely different class of vanish that's constant" makes it easier to reason about.

**Pull request merged**